lib/dialogs/logout-confirmation/index.tsx

```diff
@@ -29,15 +33,28 @@ export class LogoutConfirmation extends Component<Props> {
     exportZipArchive(notes).then(closeDialog);
   };
 
+  triggerContinueAction = () => {
+    const { action, continueAction } = this.props;
+    continueAction(action);
+  };
+
   render() {
-    const { closeDialog, notes, reallyLogout } = this.props;
+    const { action, closeDialog, notes } = this.props;
+
+    const actionDescription =
+      action === 'REALLY_LOGOUT'
+        ? 'Logging out will delete any unsynchronized notes.'
+        : 'Closing the app with unsynchronized notes could cause data loss.';
```

Comment:
the way we're passing in the action name and then making decisions based on that inside of the component feels like what I call _slanted abstractions_ where we have one concern split across separate layers of the app. in this case I feel like the confirmation dialog shouldn't have to know about the Redux actions and we shouldn't have to create a hollow function to reconstruct a _real_ Redux action based on the _partial_ action type we pass in.

Can we think of ways to reuse this component without diving in and doing surgery on it? I think I see two points of change based on the role for which we're using the dialog: labeling the description at the top; and providing a _continuation_ action for when we're done.

In the first case I wonder if we even need to have different descriptions? We _could_ pass them into the component but also would something like this work for both cases?

> If you continue and leave while there are unsynchronized changes there's a risk those changes will be lost.

Seems like something of that nature could cover both cases and the user will know the context from their interactions that they are closing vs. logging-out.

Concerning the button I'd rather see us pass in an actual action that this component doesn't know about than to pass in an action name, coupling this component to the details of our Redux system and the expectations of the parent rendering this component.

```ts
type OwnProps {
	finisher: () => any;
}

…

<button … onCLick={this.props.finisher}>
```

To that point we still have labeling on the button that's specific. Again, can we return control back up the parent? I don't know if this is the right solution, but we can have the parent supply the button itself and even avoid the question of the action.

```ts
<LogoutConfirmation key="logout-confirmation">
	({hasPendingChanges, className}) => (
		<button className={className} onClick={this.props.reallyLogOut}>
			{hasPendingChanges ? "Log out" : "Safely log out"}
		</button>
	)
</LogoutConfirmation>

<LogoutConfirmation key="close-confirmation">
	({hasPendingChanges, className}) => (
		<button className={className} onClick={this.props.reallyLogOut}>
			{hasPendingChanges ? "Close window" : "Safely close window"}
		</button>
	)
</LogoutConfirmation>
```

Reply:
I agree with everything here except that we should supply the button to the dialog. I think it is cleaner to pass the redux action to the dialog and let the dialog control the UI.